### PR TITLE
raftstore: Suppress EntryCache's growing when a TiKV is down.

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

## What have you changed? (mandatory)

When a TiKV is down, other TiKV's `EntryCache` might not be cleaned up, which is potential to cause OOM.
This PR is trying to quickly fix this issue. A peer will be considered down if it didn't response heartbeat in the past `heartbeat_interval + 10` seconds. And while the peer is calculating the range of cache to be dropped, these down peers will be ignored.

## What are the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

By testing manually.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -2195,6 +2195,11 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
     #[cfg_attr(feature = "cargo-clippy", allow(if_same_then_else))]
     fn on_raft_gc_log_tick(&mut self, event_loop: &mut EventLoop<Self>) {
+        // As leader, we believe a peer is down if we haven't receive it's heartbeat response in the
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: `probably is down`

**Comment by MyonKeminta** [comment]
Tested on 4 TiKVs, capacity 85 GB each.
split_size = 32MB
Load data by using go-ycsb:
```
./go-ycsb run tikv \
-p "tikv.pd=$PD_ADDR" -p "tikv.type=raw" \
-p "workload=core" \
-p "fieldcount=10" \
-p "fieldlength=1024" \
-p "readproportion=0" \
-p "scanproportion=0" \
-p "insertproportion=1" \
-p "updateproportion=0" \
-p "recordcount=0" \
-p "operationcount=1000000000"
```
Kill one TiKV when the store size is about 40~50GB each.


Master: (tikv_5 is killed at about 20:37)

After: (tikv_5 is killed at about 0:37)


@zhangjinpeng1987 PTAL

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -2195,6 +2195,12 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
     #[cfg_attr(feature = "cargo-clippy", allow(if_same_then_else))]
     fn on_raft_gc_log_tick(&mut self, event_loop: &mut EventLoop<Self>) {
+        // As leader, we would not keep caches for the peers that didn't response heartbeat in the
+        // last few seconds. That happens probably because another TiKV is down. In this case if we
+        // do not clean up the cache, it may keep growing.
+        let drop_cache_duration = self.cfg.raft_heartbeat_interval() + Duration::from_secs(30);
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: Please make 30 as a constant var.

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -1626,16 +1635,27 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-                    healthy_replicated_idx = p.matched;
+                if let Some(last_heartbeat) = peer.peer_heartbeats.get(peer_id) {
+                    if alive_cache_idx > p.matched
+                        && p.matched >= truncated_idx
+                        && *last_heartbeat > cache_alive_limit
+                    {
+                        alive_cache_idx = p.matched;
+                    }
+                } else {
+                    warn!(
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: I may produce a lot of logs if a node is down and leader is transferred.

**Comment by MyonKeminta** [comment]
@zhangjinpeng1987 @BusyJay PTAL

File: src/raftstore/store/fsm/peer.rs [hunk 4]
```diff
@@ -57,6 +57,8 @@ use raftstore::store::worker::{
 };
 use raftstore::store::{util, Msg, SignificantMsg, SnapKey, SnapshotDeleter, Store, Tick};
 
+const ENTRY_CACHE_LIFE_TIME_SECS: u64 = 30;
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: Better make it configurable.

File: src/raftstore/store/fsm/peer.rs [hunk 5]
```diff
@@ -1633,16 +1642,22 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            for (peer_id, p) in peer.raft_group.raft.prs().iter() {
                 if replicated_idx > p.matched {
                     replicated_idx = p.matched;
                 }
-                if healthy_replicated_idx > p.matched && p.matched >= truncated_idx {
-                    healthy_replicated_idx = p.matched;
+                if let Some(last_heartbeat) = peer.peer_heartbeats.get(peer_id) {
+                    if alive_cache_idx > p.matched
+                        && p.matched >= truncated_idx
+                        && *last_heartbeat > cache_alive_limit
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/raftstore/store/fsm/peer.rs]: We want to find the minimal value of all alive peers here, and we consider a peer to be alive if it's last_heartbeat is after the limit. If `last_heartbeat < cache_alive_limit`, on the contrary, it means that peer might be dead and it should be ignored here.

File: src/raftstore/store/fsm/peer.rs [hunk 6]
```diff
@@ -1612,6 +1614,13 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
     #[cfg_attr(feature = "cargo-clippy", allow(if_same_then_else))]
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply src/raftstore/store/fsm/peer.rs]: At round L1680 there are a pair of if-else blocks with the same content. It must was wrote for some reason. If not, it's a better idea to fix that in another PR, not this.

**Comment by MyonKeminta** [comment]
@BusyJay @overvenus PTAL again

**Pull request merged**